tests/nn/data_parallel/test_fsdp_shared_weights.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-class-docstring
+# pylint: disable=missing-function-docstring
+
+""" Test FSDP with shared weights between wrappers. """
+
+from copy import deepcopy
+
+import pytest
+import torch
+import torch.multiprocessing as mp
+from torch.nn import Linear, Module
+from torch.optim import SGD
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
+from fairscale.utils.testing import dist_init, objects_are_equal, skip_if_single_gpu, teardown, temp_files_ctx
+
+
+class Model(Module):
+    def __init__(self, with_fsdp=False, inner_flat=False, sharing=None):
+        super().__init__()
+        self.l0 = Linear(4, 4, bias=True).cuda()
+        self.l1 = Linear(4, 4, bias=True).cuda()
+        self.l2 = Linear(4, 4, bias=True).cuda()
+        self.l3 = Linear(4, 4, bias=True).cuda()
+
+        # share the weights. the layer must have at least 1 param is that's not
+        # shared. Therefore, we have bias=True and testing either sharing the
+        # weight or the bias.
+        if sharing == "share_only_weights":
+            self.l1.weight = self.l3.weight
+        elif sharing == "share_only_bias":
+            self.l1.bias = self.l3.bias
+        else:
+            assert sharing is None or sharing == "share_none"
+
+        if with_fsdp:
+            # Shared layers much be un-flatten.
+            self.l1 = FSDP(self.l1, flatten_parameters=False)
+            self.l2 = FSDP(self.l2, flatten_parameters=inner_flat)
+            self.l3 = FSDP(self.l3, flatten_parameters=False)
+
+            if sharing in ["share_only_weights"]:
+                self.l3.append_shared_param(self.l1.module.weight)
```

Comment:
Is this a change users will have to make in their models to use this feature?

Reply:
Yes